sgeop/simplify.py

```diff
@@ -198,162 +248,175 @@ def simplify_singletons(
     Returns
     -------
-
+    geopandas.GeoDataFrame
+        The road network line data following the pairs procedure.
     """
 
-    # Get nodes from the network.
-    nodes = momepy.nx_to_gdf(momepy.node_degree(momepy.gdf_to_nx(roads)), lines=False)
-
-    # Link nodes to artifacts
-    node_idx, artifact_idx = artifacts.sindex.query(
-        nodes.buffer(0.1), predicate="intersects"
-    )
-    intersects = sparse.coo_array(
-        ([True] * len(node_idx), (node_idx, artifact_idx)),
-        shape=(len(nodes), len(artifacts)),
-        dtype=np.bool_,
-    )
-
-    # Compute number of nodes per artifact
-    artifacts["node_count"] = intersects.sum(axis=0)
+    # Extract network nodes and relate to artifacts
+    nodes, artifacts = _link_nodes_artifacts("pairs", roads, artifacts, None)
 
     # Compute number of stroke groups per artifact
     roads, _ = continuity(roads)
-    strokes, c_, e_, s_ = get_stroke_info(artifacts, roads)
-
-    artifacts["stroke_count"] = strokes
-    artifacts["C"] = c_
-    artifacts["E"] = e_
-    artifacts["S"] = s_
+    artifacts = _classify_strokes(artifacts, roads)
 
-    # Filter artifacts caused by non-planar intersections.
-    artifacts["non_planar"] = artifacts["stroke_count"] > artifacts["node_count"]
-    a_idx, _ = roads.sindex.query(artifacts.geometry.boundary, predicate="overlaps")
-    artifacts.loc[artifacts.index[np.unique(a_idx)], "non_planar"] = True
+    # Filter artifacts caused by non-planar intersections
+    artifacts = _identify_non_planar(artifacts, roads)
 
-    artifacts["non_planar_cluster"] = artifacts.apply(
-        lambda x: sum(artifacts.loc[artifacts["comp"] == x.comp]["non_planar"]), axis=1
-    )
+    # Identify non-planar clusters
+    _id_np = lambda x: sum(artifacts.loc[artifacts["comp"] == x.comp]["non_planar"])  # noqa: E731
+    artifacts["non_planar_cluster"] = artifacts.apply(_id_np, axis=1)
+    # Subset non-planar clusters and planar artifacts
     np_clusters = artifacts[artifacts.non_planar_cluster > 0]
     artifacts_planar = artifacts[artifacts.non_planar_cluster == 0]
 
-    artifacts_w_info = artifacts.merge(
-        artifacts_planar.groupby("comp")[artifacts_planar.columns].apply(
-            get_solution, roads=roads
-        ),
-        left_on="comp",
-        right_index=True,
-    )
-    artifacts_under_np = np_clusters[np_clusters.non_planar_cluster == 2].dissolve(
-        "comp", as_index=False
-    )
+    # Isolate planar artifacts
+    _planar_grouped = artifacts_planar.groupby("comp")[artifacts_planar.columns]
+    _solutions = _planar_grouped.apply(get_solution, roads=roads)
+    artifacts_w_info = artifacts.merge(_solutions, left_on="comp", right_index=True)
+
+    # Isolate non-planar clusters of value 2
+    # ---> MARTIN ---> what exactly does this mean? ###################################
+    _np_clust_2 = np_clusters["non_planar_cluster"] == 2
+    # ---> MARTIN ---> what does does it mean to be 'under non-planar'? ###############
```

Comment:
You have an artifact under the highway, for example. You need to ignore the highway to solve the artifact below.